packages/core/contracts-ovm/insured-bridge/implementation/BridgePool.sol

```diff
@@ -47,32 +46,36 @@ contract BridgePool is Testable {
         address l1Token;
         address slowRelayer;
     }
-    struct Deposit {
-        DepositState depositState;
-        DepositType depositType;
+
+    // A Relay is linked to a Deposit object originating from L2
+    struct Relay {
+        RelayState relayState;
+        RelayType relayType;
         // A deposit can have both a slow and an instant relayer if a slow relay is "sped up" from slow to instant.
         // We want to store both addresses for separate payouts.
         address instantRelayer;
-        // @dev: See note above about why some Deposit params are collapsed into `RelayAncillaryDataContents`.
-        RelayAncillaryDataContents relayData;
     }
-    // Associates each deposit with a unique ID.
-    mapping(uint64 => Deposit) public deposits;
-    // If a deposit is disputed, it is removed from the `deposits` mapping and added to the `disputedDeposits` mapping.
-    // There can only be one disputed deposit per relayer for each deposit ID.
-    // @dev The mapping is `depositId-->disputer-->Deposit`
-    mapping(uint64 => mapping(address => Deposit)) public disputedDeposits;
+    // Associates each relay with a unique ancillary data hash derived from its constituent data. We need to key by the
+    // ancillary data so that the OptimisticOracle can locate relays on callbacks using only price requests' ancillary
+    // data
+    mapping(bytes32 => Relay) public relays;
+    // If a relay is disputed, it is removed from the `relays` mapping and added to the `disputedRelays` mapping.
+    // There can only be one disputed relay per relayer for each relay ancillary data.
+    // @dev The mapping is `relayAncillaryData-->disputer-->Relay`
+    mapping(bytes32 => mapping(address => Relay)) public disputedRelays;
```

Comment:
What happens if someone instant-relays, they get wrongfully disputed, and then someone else does the same relay? How do we prevent the pool from paying out twice?

I think we may need to have some sort of persistent mapping to ensure that if the deposit details match across the same request that it automatically pays out to the same instant-relayer.

Reply:
> What happens if someone instant-relays, they get wrongfully disputed, and then someone else does the same relay? How do we prevent the pool from paying out twice?

Really good catch. I think to mitigate that, we can set `newRelay.instantRelayer == disputedRelay.instantRelayer` if one exists. 

@chrismaree tagging for visibility

Reply:
Here's the path matt describes:
- Relay gets instant-relayed by someone, `recipient` receives `amount`.
- Relay gets disputed, depending on result of dispute, either the `slowRelayer` or the `disputer` gets paid out
- Another `slowRelayer2` now starts another Relay. The contract somehow knows that there is a dispute pending for this relay with an instant relayer, so the Relay becomes `type=INSTANT` and when the Relay concludes, only the relayers get paid their rewards + bonds. `recipient` receives nothing.

The tricky part is `The contract somehow knows that there is a dispute pending for this relay` because there is currently no way to locate the disputed relay when another `slowRelayer2` starts a relay.

Reply:
Exactly!

Reply:
I agree we don't need to track disputes, we can just let the OptimisticOracle pay out disputes. I ultimately settled on the following structs:

```
    // A Relay represents a an attempt to finalize a cross-chain transfer that originated on an L2 DepositBox contract
    // and can be bridged via this contract.
    enum RelayState { Uninitialized, Pending, Finalized }

    // Data from L2 deposit transaction.
    struct DepositData {
        uint64 depositTimestamp;
        uint64 maxFeePct;
        uint64 depositId;
        uint256 amount;
        address l2Sender;
        address recipient;
        address l1Token;
    }

    // A Relay is linked to a L2 Deposit.
    struct RelayData {
        RelayState relayState;
        uint256 priceRequestTime;
        uint64 proposerRewardPct;
        uint64 realizedFeePct;
        address slowRelayer;
        address instantRelayer;
    }

    // Associate deposits with pending relays. When RelayState is Uninitialized, new relay attempts can be
    // made for this deposit. Also contains information neccessary to pay out relayers on successful relay.
    mapping(bytes32 => RelayData) public relays;
    // Associates ancillary data related to relay price request with the deposit hash that the relay is attempting to 
    // fulfill. We need to key by the ancillary data so that the OptimisticOracle can locate relays on callbacks using 
    // only price requests' ancillary data. The ancillary data should contain all information required by off-chain 
    // actors (validators, DVM voters, etc.) to verify that a relay is valid.
    mapping(bytes32 => bytes32) public ancillaryDataToDepositHash;
```

This stores a forward lookup of `depositHash` to `RelayData`, so duplicate relays can't be pending for the same deposit, and a reverse lookup of `ancillaryData` to `depositHash`, so that the OO can callback into this contract and delete a pending relay--thereby freeing it up for other deposits. Any instant relayers can be carried over in the `relays` struct.

Reply:
Agreed this is a good catch. The key here is that you are setting the `proposer` in the `proposePriceFor` call so arbitration to finalize the dispute can occur directly in the OO. I think the original thinking (and the reason we stored this data) was to let this arbitration occur within the BridgePool. i.e the pool will pull rewards from the OO after the dispute, call to see who was correct and then pay out accordingly. 

This is pretty redundant seeing setting the `proposer` on the OO gives us the same arbitration mechanism without needing to store any data about the dispute within this contract.